### PR TITLE
fix open console for 4.x clusters

### LINKED ISSUE #981: open console command for clusters for OpenShift 4.x does not work
For OpenShift clusters 4.x, the api url to login is different to console URL. Hence the `OpenConsole` commands leads to the API url which is incorrect.

Resolution: The command should open the `OpenShift web-console` URL.

### PR DESCRIPTION
**Pull request by mohitsuman**
fix #981

### CODE REVIEW

File: src/openshift/cluster.ts [hunk 1]
```diff
@@ -66,17 +66,19 @@ export class Cluster extends OpenShiftItem {
         Cli.getInstance().showOutputChannel();
     }
 
-    static async openshiftConsole(context: OpenShiftObject): Promise<void> {
-        if (context) {
-            open(`${context.getName()}/console`);
+    static async openshiftConsole(): Promise<void> {
+        let consoleUrl: string;
+        const versionInfo = await Cluster.odo.execute(Command.clusterVersionDescribe(), process.cwd(), false);
```
Comment (dgolovin) [inline src/openshift/cluster.ts]: Why is it required to call `oc describe` first?

File: src/openshift/cluster.ts [hunk 2]
```diff
@@ -66,17 +66,19 @@ export class Cluster extends OpenShiftItem {
...
-            const result: OpenShiftObject[] = await Cluster.odo.getClusters();
-            if (result.length>0 && result[0].getName().startsWith('http')) {
-                open(`${result[0].getName()}/console`);
-            } else {
-                window.showErrorMessage(result[0].getName());
-            }
+            const serverUrl = await Cluster.odo.execute(Command.showServerUrl());
+            consoleUrl = `${serverUrl.stdout}/console`;
         }
+        window.showInformationMessage(`Opening console url: ${consoleUrl}`);
```
Comment (dgolovin) [inline src/openshift/cluster.ts]: No need to show the message.

File: src/openshift/cluster.ts [hunk 3]
```diff
@@ -66,17 +66,19 @@ export class Cluster extends OpenShiftItem {
...
     }
 
-    static async openshiftConsole(context: OpenShiftObject): Promise<void> {
-        if (context) {
-            open(`${context.getName()}/console`);
+    static async openshiftConsole(): Promise<void> {
+        let consoleUrl: string;
+        const versionInfo = await Cluster.odo.execute(Command.clusterVersionDescribe(), process.cwd(), false);
+        if (versionInfo.error === null) {
+            const routeObj = await Cluster.odo.execute(Command.getOpenshiftClusterRoute());
```
Comment (dgolovin) [inline src/openshift/cluster.ts]: What is output for OpenShift 3.x for this command?

File: src/openshift/cluster.ts (hunk 1 above)
Comment (mohitsuman) [reply src/openshift/cluster.ts]: Now to get to know which version of server is used, either 3.x or 4.x, there is no way from oc or odo.
The ``oc describe clusterversion` command works for 4.x server and errors out for 3.x. So the check is done for that.

For 3.x, there is no `clusterversion`.

File: src/openshift/cluster.ts (hunk 2 above)
Comment (mohitsuman) [reply src/openshift/cluster.ts]: The message was added to inform that the url is being loaded for some cases where the open in browser takes some time.

File: src/openshift/cluster.ts (hunk 3 above)
Comment (mohitsuman) [reply src/openshift/cluster.ts]: For 3.x, there is no namespace `openshift-console` and hence the route does not work here. 

For 4.x, the cluster url is generates using operator and we get the route using the namespace for `openshift-console`.

**Comment by mohitsuman** [comment]
@dgolovin After 4.2 release, I think `oc version` will have the server version also mentioned and then we can directly use that command and determine the server version.

File: src/openshift/cluster.ts (hunk 2 above)
Comment (dgolovin) [reply src/openshift/cluster.ts]: It should be shown with progress then.

File: src/openshift/cluster.ts (hunk 3 above)
Comment (dgolovin) [reply src/openshift/cluster.ts]: then clusterVersionDescribe is not needed,  just analyse getOpenshiftClusterRoute result.

Comment (mohitsuman) [reply src/openshift/cluster.ts]: But `getOpenshiftClusterRoute` for 3.x does not throw an error. It responds with empty items list. Do you want to just use the condition over the `items` object and iterate ?

File: src/openshift/cluster.ts (hunk 2 above)
Comment (mohitsuman) [reply src/openshift/cluster.ts]: Actually now I tried with progress, but it's an overkill, the open browser is sufficient for the action. Hence removing with the new commit.

File: src/openshift/cluster.ts (hunk 3 above)
Comment (mohitsuman) [reply src/openshift/cluster.ts]: @dgolovin After conversation with openshift dev team, they too suggested that `oc get clusterversion` is the correct differentiator as it works only on 4.x.

**Pull request merged**